// react-slideshow-image/lib/zoom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _tween = _interopRequireDefault(require("@tweenjs/tween.js"));

var _helpers = require("./helpers.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var Zoom =
/*#__PURE__*/
function (_Component) {
  _inherits(Zoom, _Component);

  function Zoom(props) {
    var _this;

    _classCallCheck(this, Zoom);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Zoom).call(this, props));
    _this.state = {
      index: props.defaultIndex && props.defaultIndex < props.children.length ? props.defaultIndex : 0
    };
    _this.width = 0;
    _this.timeout = null;
    _this.divsContainer = null;
    _this.wrapper = null;
    _this.setWidth = _this.setWidth.bind(_assertThisInitialized(_this));
    _this.handleResize = _this.handleResize.bind(_assertThisInitialized(_this));
    _this.navigate = _this.navigate.bind(_assertThisInitialized(_this));
    _this.preZoom = _this.preZoom.bind(_assertThisInitialized(_this));
    _this.pauseSlides = _this.pauseSlides.bind(_assertThisInitialized(_this));
    _this.startSlides = _this.startSlides.bind(_assertThisInitialized(_this));
    _this.tweenGroup = new _tween["default"].Group();
    _this.initResizeObserver = _this.initResizeObserver.bind(_assertThisInitialized(_this));
    _this.reactSlideshowWrapper = (0, _react.createRef)();
    return _this;
  }

  _createClass(Zoom, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.setWidth();
      this.play();
      this.initResizeObserver();
    }
  }, {
    key: "initResizeObserver",
    value: function initResizeObserver() {
      var _this2 = this;

      this.resizeObserver = new ResizeObserver(function (entries) {
        if (!entries) return;

        _this2.handleResize();
      });
      this.resizeObserver.observe(this.reactSlideshowWrapper.current);
    }
  }, {
    key: "play",
    value: function play() {
      var _this3 = this;

      var _this$props = this.props,
          autoplay = _this$props.autoplay,
          children = _this$props.children;
      var index = this.state.index;

      if (autoplay && children.length > 1) {
        clearTimeout(this.timeout);
        this.timeout = setTimeout(function () {
          return _this3.zoomTo(index + 1);
        }, this.props.duration);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.willUnmount = true;
      clearTimeout(this.timeout);
      this.removeResizeObserver();
    }
  }, {
    key: "removeResizeObserver",
    value: function removeResizeObserver() {
      if (this.resizeObserver && this.reactSlideshowWrapper && this.reactSlideshowWrapper.current) {
        this.resizeObserver.unobserve(this.reactSlideshowWrapper.current);
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(props) {
      if (this.props.autoplay !== props.autoplay) {
        if (this.props.autoplay) {
          this.play();
        } else {
          clearTimeout(this.timeout);
        }
      }

      if (this.props.children.length != props.children.length) {
        this.applyStyle();
        this.play();
      }
    }
  }, {
    key: "setWidth",
    value: function setWidth() {
      this.width = this.wrapper.clientWidth;
      this.applyStyle();
    }
  }, {
    key: "handleResize",
    value: function handleResize() {
      this.setWidth();
    }
  }, {
    key: "applyStyle",
    value: function applyStyle() {
      var fullwidth = this.width * this.props.children.length;
      this.divsContainer.style.width = "".concat(fullwidth, "px");

      for (var index = 0; index < this.divsContainer.children.length; index++) {
        var eachDiv = this.divsContainer.children[index];

        if (eachDiv) {
          eachDiv.style.width = "".concat(this.width, "px");
          eachDiv.style.left = "".concat(index * -this.width, "px");
        }
      }
    }
  }, {
    key: "pauseSlides",
    value: function pauseSlides() {
      if (this.props.pauseOnHover) {
        clearTimeout(this.timeout);
      }
    }
  }, {
    key: "startSlides",
    value: function startSlides() {
      var _this4 = this;

      var _this$props2 = this.props,
          pauseOnHover = _this$props2.pauseOnHover,
          autoplay = _this$props2.autoplay;

      if (pauseOnHover && autoplay) {
        this.timeout = setTimeout(function () {
          return _this4.goNext();
        }, this.props.duration);
      }
    }
  }, {
    key: "goNext",
    value: function goNext() {
      var index = this.state.index;
      var _this$props3 = this.props,
          children = _this$props3.children,
          infinite = _this$props3.infinite;

      if (!infinite && index === children.length - 1) {
        return;
      }

      this.zoomTo((index + 1) % children.length);
    }
  }, {
    key: "goBack",
    value: function goBack() {
      var index = this.state.index;
      var _this$props4 = this.props,
          children = _this$props4.children,
          infinite = _this$props4.infinite;

      if (!infinite && index === 0) {
        return;
      }

      this.zoomTo(index === 0 ? children.length - 1 : index - 1);
    }
  }, {
    key: "goTo",
    value: function goTo(index) {
      this.zoomTo(index);
    }
  }, {
    key: "navigate",
    value: function navigate(_ref) {
      var dataset = _ref.currentTarget.dataset;

      if (dataset.key != this.state.index) {
        this.goTo(parseInt(dataset.key));
      }
    }
  }, {
    key: "preZoom",
    value: function preZoom(_ref2) {
      var currentTarget = _ref2.currentTarget;

      if (currentTarget.dataset.type === 'prev') {
        this.goBack();
      } else {
        this.goNext();
      }
    }
  }, {
    key: "showIndicators",
    value: function showIndicators() {
      var _this5 = this;

      var isCustomIndicator = typeof this.props.indicators !== 'boolean';
      var className = !isCustomIndicator ? 'each-slideshow-indicator' : '';
      return _react["default"].createElement("div", {
        className: "indicators"
      }, this.props.children.map(function (each, key) {
        return _react["default"].createElement("div", {
          key: key,
          "data-key": key,
          className: "".concat(className, " ").concat(_this5.state.index === key ? 'active' : ''),
          onClick: _this5.navigate
        }, isCustomIndicator && _this5.props.indicators(key));
      }));
    }
  }, {
    key: "showPreviousArrow",
    value: function showPreviousArrow() {
      var _this$props5 = this.props,
          arrows = _this$props5.arrows,
          prevArrow = _this$props5.prevArrow,
          infinite = _this$props5.infinite;
      var className = 'custom-nav';

      if (!prevArrow) {
        className = "nav ".concat(this.state.index <= 0 && !infinite && 'disabled');
      }

      return arrows && _react["default"].createElement("div", {
        className: className,
        "data-type": "prev",
        onClick: this.preZoom
      }, prevArrow ? prevArrow : _react["default"].createElement("span", null));
    }
  }, {
    key: "showNextArrow",
    value: function showNextArrow() {
      var _this$props6 = this.props,
          arrows = _this$props6.arrows,
          nextArrow = _this$props6.nextArrow,
          infinite = _this$props6.infinite,
          children = _this$props6.children;
      var className = 'custom-nav';

      if (!nextArrow) {
        className = "nav ".concat(this.state.index === children.length - 1 && !infinite && 'disabled');
      }

      return arrows && _react["default"].createElement("div", {
        className: className,
        "data-type": "next",
        onClick: this.preZoom
      }, nextArrow ? nextArrow : _react["default"].createElement("span", null));
    }
  }, {
    key: "render",
    value: function render() {
      var _this6 = this;

      var _this$props7 = this.props,
          indicators = _this$props7.indicators,
          arrows = _this$props7.arrows,
          infinite = _this$props7.infinite,
          children = _this$props7.children;
      var index = this.state.index;
      var unhandledProps = (0, _helpers.getUnhandledProps)(Zoom.propTypes, this.props);
      return _react["default"].createElement("div", unhandledProps, _react["default"].createElement("div", {
        className: "react-slideshow-container",
        onMouseEnter: this.pauseSlides,
        onMouseLeave: this.startSlides,
        ref: this.reactSlideshowWrapper
      }, this.showPreviousArrow(), _react["default"].createElement("div", {
        className: "react-slideshow-zoom-wrapper",
        ref: function ref(_ref3) {
          return _this6.wrapper = _ref3;
        }
      }, _react["default"].createElement("div", {
        className: "zoom-wrapper",
        ref: function ref(wrap) {
          return _this6.divsContainer = wrap;
        }
      }, children.map(function (each, key) {
        return _react["default"].createElement("div", {
          style: {
            opacity: key === index ? '1' : '0',
            zIndex: key === index ? '1' : '0'
          },
          "data-index": key,
          key: key
        }, each);
      }))), this.showNextArrow()), indicators && this.showIndicators());
    }
  }, {
    key: "zoomTo",
    value: function zoomTo(newIndex) {
      var _this7 = this;

      var index = this.state.index;
      var _this$props8 = this.props,
          children = _this$props8.children,
          scale = _this$props8.scale,
          autoplay = _this$props8.autoplay,
          infinite = _this$props8.infinite,
          transitionDuration = _this$props8.transitionDuration,
          duration = _this$props8.duration,
          onChange = _this$props8.onChange;
      var existingTweens = this.tweenGroup.getAll();

      if (!existingTweens.length) {
        if (!this.divsContainer.children[newIndex]) {
          newIndex = 0;
        }

        clearTimeout(this.timeout);
        var value = {
          opacity: 0,
          scale: 1
        };

        var animate = function animate() {
          if (_this7.willUnmount) {
            _this7.tweenGroup.removeAll();

            return;
          }

          requestAnimationFrame(animate);

          _this7.tweenGroup.update();
        };

        animate();
        var tween = new _tween["default"].Tween(value, this.tweenGroup).to({
          opacity: 1,
          scale: scale
        }, transitionDuration).onUpdate(function (value) {
          _this7.divsContainer.children[newIndex].style.opacity = value.opacity;
          _this7.divsContainer.children[index].style.opacity = 1 - value.opacity;
          _this7.divsContainer.children[index].style.transform = "scale(".concat(value.scale, ")");
        }).start();
        tween.onComplete(function () {
          if (_this7.willUnmount) {
            return;
          }

          if (typeof onChange === 'function') {
            onChange(index, newIndex);
          }

          _this7.setState({
            index: newIndex
          }, function () {
            _this7.divsContainer.children[index].style.transform = "scale(1)";
          });

          if (autoplay && (infinite || newIndex < children.length - 1)) {
            clearTimeout(_this7.timeout);
            _this7.timeout = setTimeout(function () {
              _this7.zoomTo((newIndex + 1) % children.length);
            }, duration);
          }
        });
      }
    }
  }]);

  return Zoom;
}(_react.Component);

Zoom.defaultProps = {
  duration: 5000,
  transitionDuration: 1000,
  defaultIndex: 0,
  indicators: false,
  arrows: true,
  autoplay: true,
  infinite: true,
  pauseOnHover: false
};
Zoom.propTypes = {
  duration: _propTypes["default"].number,
  transitionDuration: _propTypes["default"].number,
  defaultIndex: _propTypes["default"].number,
  indicators: _propTypes["default"].oneOfType([_propTypes["default"].bool, _propTypes["default"].func]),
  scale: _propTypes["default"].number.isRequired,
  arrows: _propTypes["default"].bool,
  autoplay: _propTypes["default"].bool,
  infinite: _propTypes["default"].bool,
  onChange: _propTypes["default"].func,
  pauseOnHover: _propTypes["default"].bool,
  prevArrow: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func]),
  nextArrow: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].func])
};
var _default = Zoom;
exports["default"] = _default;
